src/commonpp/core/Gelf.cpp

```diff
@@ -0,0 +1,226 @@
+#include <boost/log/sinks/sync_frontend.hpp>
+#include <boost/log/sinks/text_ostream_backend.hpp>
+#include <boost/log/sources/global_logger_storage.hpp>
+#include <boost/log/sources/record_ostream.hpp>
+#include <boost/log/sources/severity_logger.hpp>
+#include <boost/log/trivial.hpp>
+#include <boost/log/utility/formatting_ostream.hpp>
+#include <boost/log/utility/formatting_ostream_fwd.hpp>
+#include <boost/log/utility/setup/common_attributes.hpp>
+#include <boost/log/utility/setup/formatter_parser.hpp>
+#include <iostream>
+
+#include <iostream>
+#include <random>
+#include <stdint.h>
+#include <string>
+
+namespace logging = boost::log;
+namespace sinks = logging::sinks;
+namespace expr = logging::expressions;
+
+namespace commonpp
+{
+namespace core
+{
+
+class GelfFormatter
+{
+    std::string host_;
+    std::vector<std::pair<std::string, std::string>> static_fields_;
+
+public:
+    GelfFormatter(std::string host,
+                  std::vector<std::pair<std::string, std::string>> static_fields)
+    : host_(std::move(host))
+    , static_fields_(std::move(static_fields))
+    {
+        // it might be possible to "pre-render" these to a fragment of json.
+        for (auto& s : static_fields_)
+        {
+            s.first = commonpp::string::escape_json_string(s.first);
+            s.second = commonpp::string::escape_json_string(s.second);
+        }
+    }
+
+    void operator()(const logging::record_view& rec,
+                    logging::formatting_ostream& strm)
+    {
+        strm << "{";
+        strm << "\"version\" : \"1.1\",";
+        strm << "\"host\" : \"" << host_ << "\",";
+        for (auto s : static_fields_)
+        {
+            strm << "\"" << s.first << "\" : \"" << s.second << "\",";
+        }
+
+        auto severity = rec[Severity];
+        if (!severity.empty())
+        {
+            strm << "\"level\" : " << to_syslog_level(*severity) << ",";
+        }
+        strm << "\"short_message\" : \""
+             << commonpp::string::escape_json_string(*rec[expr::smessage])
+             << "\"";
+        strm << "}";
+    }
+};
+
+class GelfUDPBackend
+    : public sinks::basic_formatted_sink_backend<char, sinks::synchronized_feeding>
+{
+    boost::asio::io_service io_service_;
+    std::unique_ptr<commonpp::net::UDPConnection> conn_;
+    std::string host_;
+    std::string port_;
+    std::mt19937_64 rng_;
+
+    // to avoid spamming cerr with logs about a bad send()/connect(), only
+    // log every SOCKET_ERROR_LOG_INTERVAL errors
+    static constexpr int SOCKET_ERROR_LOG_INTERVAL = 100;
+    static constexpr size_t MAX_PACKET_SIZE = 1000;
+    static constexpr size_t HEADER_OVERHEAD = 12;
+
+public:
+    using base =
+        sinks::basic_formatted_sink_backend<char, sinks::synchronized_feeding>;
+
+    GelfUDPBackend(std::string host, std::string port)
+    : base()
+    , host_(std::move(host))
+    , port_(std::move(port))
+    , rng_(std::random_device()())
```

Comment:
hum, I would have had a std::random_device along with the rng_ and I would have given the reference on it, however I don't think it makes a big difference

Reply:
I think I'll change it to a member. It works, but is a bit unusual and could be confusing.